pkg/apis/core/helper/helpers.go

```diff
@@ -73,14 +73,21 @@ func DetermineLatestMachineImageVersions(images []core.MachineImage) (map[string
 	return resultMapVersions, nil
 }
 
-// DetermineLatestMachineImageVersion determines the latest MachineImageVersion from a slice of MachineImageVersion
-// when filterPreviewVersions is set, versions with classification preview are not considered
+// DetermineLatestMachineImageVersion determines the latest MachineImageVersion from a slice of MachineImageVersion.
+// When filterPreviewVersions is set, versions with classification preview are not considered.
+// It will prefer older but non-deprecated versions over newer but deprecated versions.
 func DetermineLatestMachineImageVersion(versions []core.MachineImageVersion, filterPreviewVersions bool) (core.MachineImageVersion, error) {
-	latestVersion, err := DetermineLatestExpirableVersion(ToExpirableVersions(versions), filterPreviewVersions)
+	latestVersion, latestNonDeprecatedVersion, err := DetermineLatestExpirableVersion(ToExpirableVersions(versions), filterPreviewVersions)
 	if err != nil {
 		return core.MachineImageVersion{}, err
 	}
 
+	for _, version := range versions {
+		if version.Version == latestNonDeprecatedVersion.Version {
+			return version, nil
+		}
```

Comment:
Merge this into below `for` loop instead of looping again?

Reply:
Then it will be again possible `deprecated` version to be preferred over `supported` one. 
I have added a test to catch this.

Reply:
Basically, this is possible when the `deprecated` version is earlier in the `versions` slice, so it will be visited first and returned immediately.

Reply:
Understood, makes sense, thanks!

Reply:
Maybe you can add a comment about this to prevent that somebody refactors this unintentionally in the future :)

Reply:
Done with https://github.com/gardener/gardener/pull/4327/commits/16d44eed2fcb00d135cdb58754aee79df6e051ac.